### PR TITLE
EnC: Track documents that do not match the debuggee

### PR DESCRIPTION
**Pull request by tmat**
The current Project System design doesn't allow us to capture a solution snapshot at the start of a debug session that would reflect the exact content of the source files on disk the assembly was built with (https://github.com/dotnet/project-system/issues/5457). Roslyn relies on file system watches to detect changes in the source files. Events from the watcher may be handled by Roslyn at any point in time after the debugging started or not at all, since the OS does not guarantee the events to be triggered.

The PR adds a source file content checksum validation before the EnC service starts computation that depends on the content of a document at the start of the debug session. The validation compares the checksum stored in the PDB by the compiler with the current content of the file on disk. Documents are considered out-of-sync until the EnC service observes file content with matching checksum.

Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/970821
Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/992537

TODO:
To fix UWP apps [899910](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/899910) we would need to implement https://github.com/dotnet/roslyn/issues/38954. UWP generates an `.g.cs` file with an empty class, so the file is not listed in the PDB and also when launched for the first time the document that's added to the workspace is empty. Since there is no checksum to compare to the EnC service considers the empty document a baseline document and then reports rude edit when the file gets updated to the empty class.

### CODE REVIEW

**Comment by tmat** [comment]
@jasonmalinowski @ivanbasov PTAL

**Comment by tmat** [comment]
@dotnet/roslyn-compiler For compiler changes - simple refactoring so that we can share source hash algorithms with the IDE.

File: src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        /// 
+        /// Since this notification is currently not available this map tracks document snapshots of documents whose
+        /// content matches exactly to the the source file used to compile the binary currently loaded in the debuggee,
+        /// 
+        /// When EnC system analyzes document content it queries its in-sync state and adapts its behavior accordingly. 
+        /// E.g. when EnC diagnostic analyzer analyzes a document for Rude Edits it ignores out-of-sync documents to
+        /// avoid reporting false negatives. When the debugger attempts to apply changes on "continue" the EnC service
+        /// blocks the operation if there are any out-of-sync documents present.
+        /// 
+        /// Whenever we receive the file-watcher event ... TODO
```
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs]: >TODO [](start = 59, length = 4)

TODO

File: src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs [hunk 2]
```diff
@@ -188,20 +190,47 @@ private async Task TrackActiveSpansAsync()
             {
                 try
                 {
-                    var baseActiveStatements = await _editSession.BaseActiveStatements.GetValueAsync(_editSession.CancellationToken).ConfigureAwait(false);
+                    var cancellationToken = _editSession.CancellationToken;
+                    var baseActiveStatements = await _editSession.BaseActiveStatements.GetValueAsync(cancellationToken).ConfigureAwait(false);
+                    var lastCommittedSolution = _editSession.DebuggingSession.LastCommittedSolution;
+                    var currentSolution = _editSession.DebuggingSession.Workspace.CurrentSolution;
+                    var activeSpansToTrack = ArrayBuilder<(Document, Document, ITextSnapshot, ImmutableArray<ActiveStatement>)>.GetInstance();
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs]: Want to name the tuple members since you've got two Documents there?

File: src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs [hunk 3]
```diff
@@ -188,20 +190,47 @@ private async Task TrackActiveSpansAsync()
...
+                            continue;
+                        }
+
+                        if (!TryGetSnapshot(document, out var snapshot))
+                        {
+                            // Document is not open in an editor or a corresponding snapshot doesn't exist anymore.
+                            continue;
+                        }
+
+                        activeSpansToTrack.Add((baseDocument, document, snapshot, documentActiveStatements));
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs]: Comment might be needed: why do we collect them all and then do them under the lock once we're done? The old code didn't and not obvious why the change.

File: src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs [hunk 4]
```diff
@@ -232,19 +262,25 @@ private async Task TrackActiveSpansAsync()
...
+                        SetTrackingSpansNoLock(baseDocument.Id, null);
 
-                        // fire and forget:
-                        _ = RefreshTrackingSpansAsync(document, snapshot);
+                        // fire and forget on a background thread:
+                        try
+                        {
+                            _ = Task.Run(() => RefreshTrackingSpansAsync(baseDocument, document, snapshot), _editSession.CancellationToken);
+                        }
+                        catch (TaskCanceledException)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs]: Ditto: what's being thrown here? This will return a task that might be cancelled, since you're not observing it it doesn't matter.

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 5]
```diff
@@ -251,110 +237,139 @@ private LinePositionSpan GetUpToDateSpan(ActiveStatementDebugInfo activeStatemen
...
+                        exceptionRegions = default;
+                    }
 
                     builder[activeStatement.Ordinal] = new ActiveStatementExceptionRegions(exceptionRegions, isCovered);
                 }
 
-                return builder.ToImmutable();
+                var result = builder.ToImmutable();
+
+                // Only cache results if no active statements are in out-of-sync documents.
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs]: What happens if the document later comes back in sync, since OutOfSync is (theoretically) transient?

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 6]
```diff
@@ -251,110 +237,139 @@ private LinePositionSpan GetUpToDateSpan(ActiveStatementDebugInfo activeStatemen
...
             {
                 return ImmutableArray<ActiveStatementExceptionRegions>.Empty;
             }
         }
 
-        private ImmutableArray<(DocumentId DocumentId, AsyncLazy<DocumentAnalysisResults> Results)> GetChangedDocumentsAnalyses(Project baseProject, Project project)
+        private async Task<(ImmutableArray<(Document Document, AsyncLazy<DocumentAnalysisResults> Results)>, ImmutableArray<Diagnostic> Diagnostics)> GetChangedDocumentsAnalysesAsync(
+            Project baseProject, Project project, CancellationToken cancellationToken)
         {
+            var changedDocuments = ArrayBuilder<(Document Old, Document New)>.GetInstance();
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs]: Don't we do camelCase for tuple members?

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 7]
```diff
@@ -251,110 +237,139 @@ private LinePositionSpan GetUpToDateSpan(ActiveStatementDebugInfo activeStatemen
...
+            var outOfSyncDiagnostics = ArrayBuilder<Diagnostic>.GetInstance();
+
             var changes = project.GetChanges(baseProject);
+            foreach (var documentId in changes.GetChangedDocuments())
+            {
+                var document = project.GetDocument(documentId);
+                if (EditAndContinueWorkspaceService.IsDesignTimeOnlyDocument(document))
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs]: Should GetDocumentAndStateAsync just return DocumentState.DesignTimeOnly for design time only docs returned above? Otherwise this looks like there's two concepts of "DesignTimeOnly" which are different?

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 8]
```diff
@@ -251,110 +237,139 @@ private LinePositionSpan GetUpToDateSpan(ActiveStatementDebugInfo activeStatemen
...
+                    case CommittedSolution.DocumentState.DesignTimeOnly:
+                        continue;
 
-            if (changedDocuments.Length == 0)
-            {
-                return ImmutableArray<(DocumentId, AsyncLazy<DocumentAnalysisResults>)>.Empty;
-            }
+                    case CommittedSolution.DocumentState.OutOfSync:
+                        var descriptor = EditAndContinueDiagnosticDescriptors.GetDescriptor(EditAndContinueErrorCode.DocumentIsOutOfSyncWithDebuggee);
+                        outOfSyncDiagnostics.Add(Diagnostic.Create(descriptor, Location.None, new[] { document.FilePath }));
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs]: Should we tag the file in the location somehow?

File: src/EditorFeatures/Test/EditAndContinue/EditSessionActiveStatementsTests.cs [hunk 9]
```diff
@@ -461,6 +478,82 @@ static void F2()
...
+                },
+                threads: new[] { ImmutableArray.Create(thread1) });
+
+            var validator = new Validator(markedSource, activeStatements, initialState: CommittedSolution.DocumentState.OutOfSync);
+            var baseActiveStatementMap = await validator.EditSession.BaseActiveStatements.GetValueAsync(CancellationToken.None).ConfigureAwait(false);
+            var docs = validator.GetDocumentIds();
+
+            // Active Statements - available in out-of-sync documents, as they reflect the state of the debuggee and not the base document content
+
+            Assert.Equal(1, baseActiveStatementMap.DocumentMap.Count);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/EditAndContinue/EditSessionActiveStatementsTests.cs]:
```suggestion
            Assert.Single(baseActiveStatementMap.DocumentMap);
```

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs (hunk 5 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditSession.cs]: We do not cache the result until all documents that contain active statements are in-sync with debuggee. Until then the method returns a different result each time with blanks for out-of-sync docs. The callers handle not having exception handler for out-of-sync doc.

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs (hunk 7 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditSession.cs]: There are two concepts for now. We can't 100% rely on the PDB due to https://github.com/dotnet/roslyn/issues/38954. Once it's implemented we should be able to only check the PDB.

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs (hunk 6 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditSession.cs]: I don't think there is a common convention.

File: src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs [hunk 10]
```diff
@@ -102,25 +105,22 @@ public ActiveStatementTrackingSpan(ITrackingSpan trackingSpan, ActiveStatementFl
...
-                editSession.BaseSolution.Workspace.DocumentOpened += DocumentOpened;
+                editSession.DebuggingSession.Workspace.DocumentOpened += DocumentOpened;
 
                 // fire and forget on a background thread:
                 try
                 {
-                    Task.Run(TrackActiveSpansAsync, _editSession.CancellationToken);
+                    _ = Task.Run(TrackActiveSpansAsync, _editSession.CancellationToken);
                 }
                 catch (TaskCanceledException)
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/EditorFeatures/Core/Implementation/EditAndContinue/ActiveStatementTrackingService.cs]: Upon an inspection of Task.Run it seems that this is indeed unnecessary.
I'll leave it as is and follow up in another PR.
